phoenix-core/src/main/java/org/apache/phoenix/expression/function/SignFunction.java

```diff
@@ -0,0 +1,77 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.phoenix.expression.function;
+
+import java.math.BigDecimal;
+import java.sql.SQLException;
+import java.util.List;
+
+import org.apache.hadoop.hbase.io.ImmutableBytesWritable;
+import org.apache.phoenix.expression.Expression;
+import org.apache.phoenix.parse.FunctionParseNode.Argument;
+import org.apache.phoenix.parse.FunctionParseNode.BuiltInFunction;
+import org.apache.phoenix.schema.tuple.Tuple;
+import org.apache.phoenix.schema.types.PDataType;
+import org.apache.phoenix.schema.types.PDecimal;
+import org.apache.phoenix.schema.types.PInteger;
+
+/**
+ *
+ * Base class for built-in SIGN function.
+ *
+ * @since 4.3.0
+ *
+ */
+@BuiltInFunction(name = SignFunction.NAME,
+                 args = {
+                        @Argument(allowedTypes={PDecimal.class})
+                        }
+                )
+public class SignFunction extends ScalarFunction {
```

Comment:
Nice work! I think performance of this can be improved by not having to create a BigDecimal in evaluate. One way would be to introduce a PDataType.getSign() method that would work directly from the numeric type of the child (i.e. long, int, short, byte). You'd need to maybe throw in cases where getSign doesn't make sense, like with VARCHAR or CHAR. Another way would be to introduce a PNumericType and introduce the getSign method there. Then reparent all the numeric PDataType to this new class. You'd know you could cast the childExpr.getDataType() method to PNumericType because we guarantee that at compile time.